New tag field correctly clears when new tag added.

ExperimentMonty opened the pull request:
When you add a new tag, the code looks like it should clear out that text field, but it doesn't, and a user might not realize the tag was added and end up adding multiple identical tags that need to be cleaned up later.  Now it correctly clears the new tag text field after a tag is added.

ExperimentMonty committed:
New tag field correctly clears when new tag added.

gisikw commented:
This works for me, but it feels hacky. It looks like `$scope !== this` within that callback. I'm nervous this could break the ng-model binding in the future (though it doesn't seem to at the moment).

@ExperimentMonty Do you have a sense of what's causing the broken 2-way data binding to begin with?

gisikw commented:
OH! I see what's going on. Problem is that we're changing a top-level `$scope` property, so the `ng-model` reference is broken.

So a canonical Angular solution here would be something like:

``` js
// inside the controller function
$scope._newTag = {name:''};
/*... */
$scope.createTag = function(name) {
  User.user.ops.addTag({body:{name:name, id:Shared.uuid()}});
  $scope._newTag.name = '';
}
```

Then simply update the reference in website/views/main/filters.jade to reference `_newTag.name`, rather than just `_newTag`.

@ExperimentMonty does that make sense?

crookedneighbor commented:
@gisikw Good suggestions. @ExperimentMonty do you feel up to trying that out?

ExperimentMonty commented:
Sounds good to me. It's my first time with angular, so I figured I'd probably do something non-idiomatic.  I'll test that out in the morning and push a new commit if it works out.

ExperimentMonty committed:
Changed new tag code to be more angular-idiomatic.

ExperimentMonty commented:
I think I added in the changes you wanted correctly, but then it was adding an [Object object] as the tag text, so I changed the createTag function a bit.  It all works as intended, not sure how idiomatically angular it is.

gisikw commented:
@ExperimentMonty Heya! Looks good - sorry, I should have clarified about that. With ng-model working correctly, we actually don't need to pass an argument to the function, since Angular already handles the data-binding between the DOM and the JavaScript object. Lemme make some minor notes on the commit.

gisikw commented:
Oh, and it looks like there's a spec that needs to be updated to reflect this change. If you look at `test/spec/filtersCtrlSpec.js`, you'll wanna change the "creates a tag" spec to look like:

``` js
  it('creates a tag', function(){
    scope._newTag = {name:'tagName'}
    scope.createTag();
    expect(user.tags).to.have.length(1);
    expect(user.tags[0].name).to.eql('tagName');
    expect(user.tags[0]).to.have.property('id');
  });
```

Cheers!

ExperimentMonty committed:
Updates new tag code and test spec for new tags.

ExperimentMonty commented:
Okay, new build has those changes and is passing tests.

ExperimentMonty committed:
Forgot to change part of the new tag test.

crookedneighbor commented on website/views/main/filters.jade:
```diff
@@ -21,7 +21,7 @@
         li.filters-edit(ng-show='_editing')
           form.hrpg-input-group
             input(type='text', ng-model='_newTag', placeholder=env.t('newTag'))
```
This should be `ng-model='_newTag.name'`, I believe.

gisikw replied:
:+1:

ExperimentMonty replied:
Totally correct, I made some poor git decisions and somehow lost those lines by the time I pushed my changes.

ExperimentMonty replied:
And I had lost a line from the filtersCtrl section too, the $scope._newTag = {name:''}; line. :-1:

ExperimentMonty committed:
Fixed errors caused by poor git-fu.

Forgot I was on a detached head, made a commit, and failed hard at
cherry-picking the commit back to the correct branch.

ExperimentMonty committed:
Fixes more bad git-fu. New tag creation works now.

ExperimentMonty commented:
Okay, added that last fix in. Sorry for drawing out this simple fix!

gisikw commented:
Looks great to me :+1:

Alys commented:
@ExperimentMonty Thank you! This is a nice little fix. Can you tell me your user ID (Settings -> API) so I can give you a contributor tier?

@gisikw Thank you for your help here! It was very useful.

ExperimentMonty commented:
@Alys Thank you very much!  My user ID is 34cda205-f8a1-416a-9cf6-4b75bd46a874

@gisikw Thanks for your help also!

Alys commented:
Tier 1 awarded. Congratulations @ExperimentMonty!
